sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+    pub fn size_in_bytes(&self) -> usize {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value.num_bytes() as usize
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value.len()
+        }
+    }
+
+    /// Returns a big-endian byte vector representation
+    /// of the absolute value of `self`
+    pub fn to_vec(&self) -> Vec<u8> {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            if self.value.num_bytes() == 0 {
+                return vec![0u8];
+            }
+            self.value.to_vec()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value.clone()
+        }
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(u32_val: u32) -> Self {
+        if u32_val == 0 {
+            Self::new()
+        } else {
+            #[cfg(not(target_arch = "bpf"))]
+            {
+                let bn = openssl::bn::BigNum::from_u32(u32_val).unwrap();
+                Self { value: bn }
+            }
+            #[cfg(target_arch = "bpf")]
+            {
+                extern "C" {
+                    fn sol_bignum_from_u32(bn_syscall_result: *mut u64, u32_value: u64) -> u64;
+                }
+                let mut value = Vec::<u8>::with_capacity(4);
+                let mut bn_syscall_result = {
+                    FfiBigNumber {
+                        data: value.as_mut_ptr() as *mut _ as u64,
+                        data_len: 4u64,
+                        is_negative: false,
+                    }
+                };
+                unsafe {
+                    sol_bignum_from_u32(
+                        &mut bn_syscall_result as *mut _ as *mut u64,
+                        u32_val as u64,
+                    );
+                    value.set_len(bn_syscall_result.data_len as usize);
+                    Self {
+                        is_negative: bn_syscall_result.is_negative,
+                        value,
+                    }
+                }
+            }
+        }
+    }
+
+    /// Returns a BigNumber from a decimal string
+    pub fn from_dec_str(string: &str) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            let bn = openssl::bn::BigNum::from_dec_str(string).unwrap();
+            Self { value: bn }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_from_dec_str(
+                    in_dec_str_ptr: *const u64,
+                    in_size: u64,
+                    bn_syscall_result: *mut u64,
+                ) -> u64;
+            }
+            let mut value = Vec::<u8>::with_capacity(string.len());
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: string.len() as u64,
+                    is_negative: false,
```

Comment:
I guess the string cannot contain a sign?  Might want to add that this is only good for positive numbers to the function comment

Reply:
@jackcmay No, you can pass a "-12345" to `from_dec_str(...)` and it will work as expected